### PR TITLE
fix: parse empty array of structs into empty list

### PR DESCRIPTION
**Pull request by fubuloubu**
### What I did

Avoid trying to create a struct value when the returndata is empty if the return data is expeted to be a dynamic array of structs

fixes: #1231

fixes: APE-490

### How I did it

Found two conditions that needed to be fixed:
1. If there is no data, you cannot parse out any components for a struct, so return `None` early (avoids `TypeError` per #1231)
2. If the type is an array of structs, and you get a `None` value, then it should be an empty list instead.

### How to verify it

Not sure if this was the right approach to solving the problem, it seems weird to me that `_create_struct` would not be handled better by `parse` upstream of the second change I had to make to handle that condition

### Checklist

- [x] All changes are completed
- [ ] New test cases have been added
- ~~[ ] Documentation has been updated~~

### CODE REVIEW

**Comment by fubuloubu** [comment]
Needs tests for both tuple of dynarray of structs and dynarray of structs

File: src/ape/utils/abi.py [hunk 1]
```diff
@@ -111,6 +111,10 @@ def parse(self, output_types: List[ABIType], values: Union[List, Tuple]) -> Any:
                         }
                         item_type = ABIType.parse_obj(item_type_data)
                         parsed_item = self.parse([item_type], [value])
+                        # If it's an empty array of structs, replace `None` with empty list
+                        if output_type.type == "tuple[]" and parsed_item is None:
```
Comment (unparalleled-js) [inline src/ape/utils/abi.py:115]: What about fixed size, like `tuple[2]`?

**Review by unparalleled-js — COMMENTED** [review]
We could use a test for this. Let me know if need help or would like me to do it.

File: src/ape/utils/abi.py (hunk 1 above)
Comment (fubuloubu) [reply src/ape/utils/abi.py:115]: A lot more permutations to test, some examples:
```vyper
struct Foo:
    a: address
    b: bytes32
    c: uint256

...
def bar1() -> Foo:
...
def bar2() -> DynArray[Foo, N]:
...
def bar3() -> (DynArray[Foo, N], DynArray[Foo, N]):
...
def bar4() -> (Foo[N], DynArray[Foo, N]):
...
def bar5() -> DynArray[Foo[M], N]:
```

File: src/ape/utils/abi.py [hunk 2]
```diff
@@ -132,6 +136,9 @@ def _create_struct(self, out_abi: ABIType, out_value) -> Optional[Any]:
             name = out_abi.name or self.default_name
 
         components = self._parse_components(out_abi.components, out_value[0])
+        if not components:
+            return None  # Empty struct
```
Comment (unparalleled-js) [inline src/ape/utils/abi.py:140]: When return `empty(MyStruct)` in Vyper, it gives me a components like this: 
```
['0x0000000000000000000000000000000000000000', HexBytes('0x0000000000000000000000000000000000000000000000000000000000000000')]
```

Is `empty(MyStruct)` not what you mean by `# Empty struct` here?

Comment (unparalleled-js) [reply src/ape/utils/abi.py:140]: Were you able to hit this return statement in your example?

Comment (fubuloubu) [reply src/ape/utils/abi.py:140]: `empty(MyStruct)` means "a `MyStruct` with all default (empty) values", so it actually does produce a struct (which is perhaps confusing)

Comment (fubuloubu) [reply src/ape/utils/abi.py:140]: An "empty struct" is a struct w/o any components, e.g. `struct MyStruct: pass`, which I don't think is valid in Vyper (and maybe Solidity)

Comment (fubuloubu) [reply src/ape/utils/abi.py:140]: Perhaps this should raise instead

Comment (unparalleled-js) [reply src/ape/utils/abi.py]: It is needed for empty tuples actually, but I adjusted it a little.

File: src/ape/utils/abi.py (hunk 1 above)
Comment (unparalleled-js) [reply src/ape/utils/abi.py:115]: I am adding these now and adjusting

Comment (unparalleled-js) [reply src/ape/utils/abi.py:115]: I added a bunch more tests cases that I think align with these.
Let me know if you think anything is missing yet!

File: tests/functional/test_contract_instance.py [hunk 3]
```diff
@@ -194,7 +195,35 @@ def test_nested_structs_in_tuples(contract_instance, sender, chain):
     assert is_checksum_address(struct_2.t.a)
 
 
-def test_vyper_structs_with_array(vyper_contract_instance, sender):
+def test_get_empty_dyn_array_of_structs(contract_instance):
+    actual = contract_instance.getEmptyDynArrayOfStructs()
+    expected = None
```
Comment (unparalleled-js) [inline tests/functional/test_contract_instance.py:200]: I think this should actually be an empty list.

Comment (unparalleled-js) [reply tests/functional/test_contract_instance.py]: k it list now

**Review by fubuloubu — COMMENTED** [review]
Approved

**Review by unparalleled-js — APPROVED** [review]

**Pull request merged**